Improve Self-Hosting Guide (#308)
dbos-inc/dbos-docs: docs/self-hosting/self-hosting.md

---
sidebar_position: 1
title: Self-Hosting DBOS Apps
---

Because DBOS is implemented in lightweight TypeScript and Python libraries, you can run your DBOS application anywhere as long as it has a Postgres server to connect to.

> This guide provides information on operating a self-hosted DBOS application.

Reviewer: I'd expand on the concepts first. It is important people understand what are the DBOS components which require special management (spoiler: versioned workflows & their recovery)

- `you can run your DBOS application anywhere`: a DBOS app is a normal Python / node process that can be managed by any cluster manager (e.g., your local machine, kubernetes)
- What is different about a DBOS app? It has (versioned) workflows and recovery processes.

-> explain workflow versioning
-> explain recovery process
-> Then go over the difference between 1 host and multiple hosts

- **Endpoint**: `/dbos-healthz`
- **HTTP Method**: GET
- **Description**: Performs a health check on the application.
- **Response**:
  - **Status Code**: 200 OK if the system is healthy; otherwise, appropriate error codes.

### Workflow Recovery

- **Endpoint**: `/dbos-workflow-recovery`
- **Method**: POST
- **Description**: Recover all pending workflows associated with input [executor IDs](#managing-workflow-recovery). Returns the IDs of all workflows queued for recovery.
- **Request Body Format**: JSON list of executors whose pending workflows to recover.
  - **Example**:
    ```json
    ["executor-id-1", "executor-id-2", "..."]
    ```
- **Response**:
  - **Status Code**: 200 OK on successful recovery initiation; otherwise, appropriate error codes.
  - **Body Format**: JSON list of the IDs of workflows queued for recovery.
  - **Example**:
    ```json
    ["workflow-uuid-1", "workflow-uuid-2", "..."]
    ```

### Deactivate

- **Endpoint**: `/deactivate`
- **Method**: GET
- **Description**: Deactivate an executor. A deactivated executor may complete active workflows and recover `PENDING` workflows, but may not start new workflows or dequeue workflows.
- **Response**:

>   - **Status Code**: 200 OK if the request succeeeded; otherwise, appropriate error codes.

Reviewer: What about the workflow management endpoints? Want to add them later once the full integration is done?

I'll add the queue metadata docs when its done -- I'd not advocate for documenting the perf endpoint because it can be unresponsive if the app is overloaded.

Author: These are the important endpoints. We can add documentation for the others later, maybe when they're done.

---
sidebar_position: 1
title: Self-Hosting DBOS Apps
---

Because DBOS is implemented in lightweight TypeScript and Python libraries, you can **run your DBOS application anywhere** as long as it has a Postgres server to connect to.
DBOS runs entirely inside your process, executing your workflows and recording their state in Postgres.
By default, when your process restarts, DBOS recovers all its pending workflows.
However, care must be taken when upgrading your application's code or when managing many DBOS processes at scale to ensure all your workflows are efficiently recovered.

This guide will describe best practices for self-hosting an application that uses the open-source DBOS libraries, explaining how to manage workflow recovery both when operating on a single server and when operating at scale.

## Self-Hosting On A Single Server

> Self-hosting a DBOS application on a single server is simple: each time you restart your application's process, it recovers all workflows that were executing before the restart (all `PENDING` workflows).

Reviewer: ```suggestion
Self-hosting a DBOS application on a single server is simple: each time you restart your application's process, it recovers all workflows that were executing before the restart -- all `PENDING` workflows -- if the workflow code did not change (see [Managing Application versions](#managing-application-versions) bellow .
```

## Self-Hosting on Multiple Servers

When self-hosting in a distributed setting, it is important to manage workflow recovery so that when an executor crashes, restarts, or is shut down, its workflows are recovered.
You should assign each executor running a DBOS application an executor ID by setting the `DBOS__VMID` environment variable.
Each workflow is tagged with the ID of the executor that started it.
When an application with an executor ID restarts, it only recovers pending workflows assigned to that executor ID.
You can also instruct your executor to recover workflows assigned to other executor IDs through the [workflow recovery endpoint of the admin API](#workflow-recovery).

## Managing Application versions

When self-hosting, it is important to be careful when upgrading your application's code.

> When DBOS is launched, it computes an "application version" from a checksum of the code in your application's workflows (you can override this version through the `DBOS__APPVERSION` environment variable).

Reviewer: ```suggestion
When DBOS is launched, it computes an "application version" from a checksum of your application's workflows, that is, their actual code. You can override this version through the `DBOS__APPVERSION` environment variable.
```

---
sidebar_position: 1
title: Self-Hosting DBOS Apps
---

Because DBOS is implemented in lightweight TypeScript and Python libraries, you can **run your DBOS application anywhere** as long as it has a Postgres server to connect to.
DBOS runs entirely inside your process, executing your workflows and recording their state in Postgres.
By default, when your process restarts, DBOS recovers all its pending workflows.
However, care must be taken when upgrading your application's code or when managing many DBOS processes at scale to ensure all your workflows are efficiently recovered.

> This guide will describe best practices for self-hosting an application that uses DBOS, explaining how to manage workflow recovery both when operating on a single server and when operating at scale.

Reviewer 2: ```suggestion
This guide will describe best practices for self-hosting an application that uses the open-source DBOS libraries, explaining how to manage workflow recovery both when operating on a single server and when operating at scale.
```

Each workflow is tagged with the version of the application that started it.
When a DBOS application starts, it does not recover workflows tagged with a different application version.
Thus, to safely recover workflows started on an older version of your code, you should start a process running that code version.

## Self-Hosting on Multiple Servers

When self-hosting in a distributed setting, it is important to manage workflow recovery so that when an executor crashes, restarts, or is shut down, its workflows are recovered.
You should assign each executor running a DBOS application an executor ID by setting the `DBOS__VMID` environment variable.
Each workflow is tagged with the ID of the executor that started it.
When an application with an executor ID restarts, it only recovers pending workflows assigned to that executor ID.
You can also instruct your executor to recover workflows assigned to other executor IDs through the [workflow recovery endpoint of the admin API](#workflow-recovery).

> It is also important to be careful when upgrading your application's code.

Reviewer 2: This is a duplicate. I think we should explain the app version concept in a separate subsection before single server and multiple servers.

When DBOS is launched, it computes an "application version" from a checksum of the code in your application's workflows (you can override this version through the `DBOS__APPVERSION` environment variable).
Each workflow is tagged with the version of the application that started it.
When a DBOS application starts, it does not recover workflows tagged with a different application version.

> To safely recover workflows started on an older version of your code, you should start a process running that code version and use the [workflow recovery endpoint of the admin API](#workflow-recovery) to instruct it to recover workflows belonging to executors that ran old versions of DBOS.

Reviewer 2: Unclear why you need to have different ways to resume old versions for a single server versus multiple. We can clarify that by saying "As mentioned above, a process only recovers pending workflows assigned to its own executor ID with the same version. Therefore, start a process running the old version and instruct it to recover ...."
